backend/deepchecks_monitoring/bgtasks/object_storage_ingestor.py

```diff
@@ -0,0 +1,158 @@
+        ))
+
+        if organization_schema is None:
+            await session.commit()
+            return
+
+        await database.attach_schema_switcher_listener(
+            session=session,
+            schema_search_path=[organization_schema, 'public']
+        )
+
+        model: Model = (await session.scalar(select(Model).options(selectinload(Model.versions))
+                                             .where(Model.id == model_id)))
+        if model is None:
+            await session.commit()
+            return
+
+        # Get s3 authentication info
+        s3_data_source = (await session.scalar(select(DataSource).where(DataSource.type == 's3')))
+        if s3_data_source is None:
+            # TODO: Write error to somewhere
+            await session.commit()
+            return
+
+        access_key = s3_data_source.parameters['aws_access_key_id']
+        secret_key = s3_data_source.parameters['aws_secret_access_key']
+        s3 = boto3.client(
+            's3',
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key
+        )
+
+        s3_url = urlparse(model.s3_path)
+        model_path = Path(s3_url.path)
+        bucket = s3_url.netloc
+        new_scan_time = pdl.now()
+        # First ever scan of model - will scan all files
+        if model.s3_last_scan_time is None:
+            model_prefixes = ['']
+        # Else scan only new files since last scan (by date)
+        else:
+            model_prefixes = []
+            date = pdl.instance(model.s3_last_scan_time).date()
+            while date <= new_scan_time.date():
+                date = date.add(days=1)
+                model_prefixes.append(date.isoformat())
+
+        version: ModelVersion
+        for version in model.versions:
+            version_path = model_path / version.name
+            # If first scan of specific version - scan all files under version, else scan only new files by date
+            version_prefixes = model_prefixes if version.last_file_ingested is not None else ['']
+            for prefix in version_prefixes:
+                for df, time in ingest_prefix(s3, bucket, f'{version_path}/{prefix}', version.latest_file_time):
+                    await self.ingestion_backend.log_samples(version, df, session, organization_id, new_scan_time)
+                    version.latest_file_time = max(version.latest_file_time, time)
+
+        # Ingest labels
+        for prefix in model_prefixes:
+            for df, time in ingest_prefix(s3, bucket, f'{model_path}/labels/{prefix}', model.latest_labels_file_time):
+                await self.ingestion_backend.log_labels(model, df, session, organization_id)
+                model.latest_labels_file_time = max(model.latest_labels_file_time, time)
+
+        model.s3_last_scan_time = new_scan_time
+        await session.commit()
+        s3.close()
+
+
+def ingest_prefix(s3, bucket, prefix, last_file_time=pdl.date(1970, 1, 1)):
```

Comment:
not sure, but IMO pyarrow might facilitate this task/function

- [Listing/Reading files from S3](https://arrow.apache.org/docs/python/filesystems.html#s3)
- [Tabular Datasets](https://arrow.apache.org/docs/python/dataset.html#tabular-datasets)

Reply:
Interesting thanks! 
it doesn't really reduce a lot of code right now since iterating files is 3 rows, most of the logic is parsing the file name, comparing the dates, etc